plugins/node/opentelemetry-instrumentation-fastify/src/instrumentation.ts

```diff
@@ -0,0 +1,269 @@
+import type { FastifyReply } from 'fastify/types/reply';
+import type { FastifyRequest } from 'fastify/types/request';
+import {
+  AttributeNames,
+  FastifyNames,
+  FastifyTypes,
+} from './enums/AttributeNames';
+import type { HandlerOriginal, PluginFastifyReply } from './types';
+import {
+  endSpan,
+  safeExecuteInTheMiddleMaybePromise,
+  startSpan,
+} from './utils';
+import { VERSION } from './version';
+
+export const ANONYMOUS_NAME = 'anonymous';
+
+/** Fastify instrumentation for OpenTelemetry */
+export class FastifyInstrumentation extends InstrumentationBase {
+  constructor(config: InstrumentationConfig = {}) {
+    super(
+      '@opentelemetry/instrumentation-fastify',
+      VERSION,
+      Object.assign({}, config)
+    );
+  }
+
+  init() {
+    return [
+      new InstrumentationNodeModuleDefinition<any>(
+        'fastify',
+        ['^3.0.0'],
+        (moduleExports, moduleVersion) => {
+          this._diag.debug(`Applying patch for fastify@${moduleVersion}`);
+          return this._patchConstructor(moduleExports);
+        }
+      ),
+    ];
+  }
+
+  private _hookOnRequest() {
+    const instrumentation = this;
+    return function onRequest(
+      request: FastifyRequest,
+      reply: FastifyReply,
+      done: HookHandlerDoneFunction
+    ) {
+      if (!instrumentation.isEnabled()) {
+        return done();
+      }
+      instrumentation._wrap(reply, 'send', instrumentation._patchSend());
+
+      const rpcMetadata = getRPCMetadata(context.active());
+      const routeName = request.routerPath;
+      if (routeName && rpcMetadata?.type === RPCType.HTTP) {
+        rpcMetadata.span.setAttribute(SemanticAttributes.HTTP_ROUTE, routeName);
+        rpcMetadata.span.updateName(`${request.method} ${routeName || '/'}`);
+      }
+      done();
+    };
+  }
+
+  private _wrapHandler(
+    pluginName: string,
+    hookName: string,
+    original: (...args: unknown[]) => Promise<any> | void,
+    syncFunctionWithDone: boolean
+  ): () => Promise<any> | void {
+    const instrumentation = this;
+    return function (this: any, ...args: unknown[]): Promise<any> | void {
+      if (!instrumentation.isEnabled()) {
+        return original.apply(this, args);
+      }
+
+      const spanName = `${FastifyNames.MIDDLEWARE} - ${
+        original.name || ANONYMOUS_NAME
+      }`;
+
+      const reply = args[1] as PluginFastifyReply;
+
+      const span = startSpan(reply, instrumentation.tracer, spanName, {
+        [AttributeNames.FASTIFY_TYPE]: FastifyTypes.MIDDLEWARE,
+        [AttributeNames.PLUGIN_NAME]: pluginName,
+        [AttributeNames.HOOK_NAME]: hookName,
+      });
+
+      const origDone =
+        syncFunctionWithDone &&
+        (args[args.length - 1] as HookHandlerDoneFunction);
+      if (origDone) {
+        args[args.length - 1] = function (
+          ...doneArgs: Parameters<HookHandlerDoneFunction>
+        ) {
+          endSpan(reply);
+          origDone.apply(this, doneArgs);
```

Comment:
This done patched function will be called with the context of the hook span (or some other span which was started in the hook).
I believe we need to unset this context - e.g. capture the original context before the hook started and `context.with(origContext, origDone, this, ...doneArgs)`